controller/authorize_blackbox_test.go

```diff
@@ -0,0 +1,198 @@
+	// The request originates from Keycloak after a valid authorization by the end user.
+	refererKeycloakUrl := "https://keycloak-url.example.org/path-of-login"
+	req.Header.Add("referer", refererKeycloakUrl)
+
+	goaCtx = goa.NewContext(goa.WithAction(ctx, "AuthorizecallbackTest"), rw, req, prms)
+	callbackCtx, err := app.NewCallbackAuthorizeContext(goaCtx, req, goa.New("LoginService"))
+	require.Nil(t, err)
+
+	err = ctrl.Callback(callbackCtx)
+	require.Nil(t, err)
+
+	assert.Equal(t, 307, rw.Code) // redirect to keycloak login page.
+
+	assert.Contains(t, rw.Header().Get("Location"), redirectURI)
+
+	locationString = rw.HeaderMap["Location"][0]
+	locationUrl, err = url.Parse(locationString)
+	require.Nil(t, err)
+
+	allQueryParameters = locationUrl.Query()
+
+	assert.NotNil(t, allQueryParameters)
+	assert.NotNil(t, allQueryParameters["state"][0])
+	assert.NotNil(t, allQueryParameters["code"][0])
+}
+
+func (rest *TestAuthorizeREST) TestAuthorizeBadRequest() {
+	t := rest.T()
+
+	rw := httptest.NewRecorder()
+	u := &url.URL{
+		Path: fmt.Sprintf("/api/authorize"),
+	}
+	req, err := http.NewRequest("GET", u.String(), nil)
+	require.Nil(t, err)
+
+	redirectURI := "https://openshift.io/somepath"
+	prms := url.Values{}
+
+	prms.Add("response_type", "code")
+	prms.Add("redirect_uri", redirectURI)
+	prms.Add("client_id", "5dec5fdb-09e3-4453-b73f-5c828832b28e")
+	prms.Add("state", uuid.NewV4().String())
+
+	ctx := context.Background()
+
+	prms.Del("response_type")
+	goaCtx := goa.NewContext(goa.WithAction(ctx, "AuthorizeTest"), rw, req, prms)
+	_, err = app.NewAuthorizeAuthorizeContext(goaCtx, req, goa.New("LoginService"))
+	require.NotNil(t, err)
+
+	prms.Del("redirect_uri")
+	goaCtx = goa.NewContext(goa.WithAction(ctx, "AuthorizeTest"), rw, req, prms)
+	_, err = app.NewAuthorizeAuthorizeContext(goaCtx, req, goa.New("LoginService"))
+	require.NotNil(t, err)
+
+	prms.Del("client_id")
+	goaCtx = goa.NewContext(goa.WithAction(ctx, "AuthorizeTest"), rw, req, prms)
+	_, err = app.NewAuthorizeAuthorizeContext(goaCtx, req, goa.New("LoginService"))
+	require.NotNil(t, err)
+
+	prms.Del("state")
+	goaCtx = goa.NewContext(goa.WithAction(ctx, "AuthorizeTest"), rw, req, prms)
+	_, err = app.NewAuthorizeAuthorizeContext(goaCtx, req, goa.New("LoginService"))
+	require.NotNil(t, err)
+}
+
+func (rest *TestAuthorizeREST) TestAuthorizeCallbackBadRequest() {
+	t := rest.T()
+
+	rw := httptest.NewRecorder()
+	u := &url.URL{
+		Path: fmt.Sprintf("/api/authorize/callback"),
+	}
+	req, err := http.NewRequest("GET", u.String(), nil)
+	require.Nil(t, err)
+
+	prms := url.Values{}
+	prms.Add("code", "SOME_OAUTH2.0_CODE")
+	prms.Add("state", uuid.NewV4().String())
+
+	ctx := context.Background()
+
+	prms.Del("code")
+	goaCtx := goa.NewContext(goa.WithAction(ctx, "AuthorizeTest"), rw, req, prms)
+	_, err = app.NewAuthorizeAuthorizeContext(goaCtx, req, goa.New("LoginService"))
+	require.NotNil(t, err)
```

Comment:
The code isn't checking if a 400 Bad request has been returned. the `rw` object should the response code.

Reply:
We have made these parameters compulsory in design, so that we don't need to check.Since we have no checks, parameters will be accepted. It will result in 200. Do we want that?

Reply:
I'm not suggesting you to not do `app.NewAuthorizeAuthorizeContext`, rather I'm suggesting that you could see if the `rw` object has information about the response?

Reply:
app.NewAuthorizeAuthorizeContext is setting compulsory fields to empty strings, so we won't see bad parameter error on this api. I can add checks for empty strings.